alibi/explainers/counterfactual_tf2.py

```diff
@@ -0,0 +1,1028 @@
+#  define an object/series of functions to check if PyTorch/TF/both are installed and return an
+#  appropriate casting function.
+
+@singledispatch
+def to_numpy_arr(X: Union[tf.Tensor, tf.Variable, np.ndarray]):
+
+    supported_types = [tf.Tensor, tf.Variable, np.ndarray]
+    raise TypeError(f"Expected input object type to be one of {supported_types} but got type {type(X)}!")
+
+
+@to_numpy_arr.register
+def _(X: np.ndarray) -> np.ndarray:
+    return X
+
+
+@to_numpy_arr.register
+def _(X: tf.Tensor) -> np.ndarray:
+    return X.numpy()
+
+
+@to_numpy_arr.register
+def _(X: tf.Variable) -> np.ndarray:
+    return X.numpy()
+
+
+# TODO: ALEX: TBD I think it's best practice to define all the custom
+#  errors in one file (e.g., exceptions.py/error.py). We
+#  could standardise this across the library
+
+
+class CounterfactualError(Exception):
+    pass
+
+
+CF_SUPPORTED_DISTANCE_FUNCTIONS = ['l1']
+CF_VALID_SCALING = ['MAD']
+CF_PARAMS = ['scale_loss', 'constrain_features', 'feature_whitelist']
+
+
+def counterfactual_loss(instance: tf.Tensor,
+                        cf: tf.Variable,
+                        lam: float,
+                        feature_scale: tf.Tensor,
+                        pred_probas: Optional[tf.Variable] = None,
+                        target_probas: Optional[tf.Tensor] = None,
+                        distance_fcn: str ='l1',
+                        ) -> tf.Tensor:
+
+    # TODO: ALEX: TBD: TO SUPPORT BLACK-BOX, PYTORCH WE NEED A SIMILAR FUNCTION AND A HIGHER LEVEL ROUTINE TO
+    #  "DISPATCH" AMONG THESE FUNCTIONS
+
+    # TODO: ALEX: DOCSTRING
+
+    distance = lam * distance_loss(instance, cf, feature_scale, distance_fcn=distance_fcn)
+    if pred_probas is None or target_probas is None:
+        return distance
+    pred = pred_loss(pred_probas, target_probas)
+    return distance + pred
+
+
+def distance_loss(instance: tf.Tensor,
+                  cf: tf.Variable,
+                  feature_scale: tf.Tensor,
+                  distance_fcn: str = 'l1') -> tf.Tensor:
+
+    # TODO: ALEX: DOCSTRING
+
+    if distance_fcn not in CF_SUPPORTED_DISTANCE_FUNCTIONS:
+        raise NotImplementedError(f"Distance function {distance_fcn} is not supported!")
+
+    ax_sum = tuple(np.arange(1, len(instance.shape)))
+
+    return tf.reduce_sum(tf.abs(cf - instance) / feature_scale, axis=ax_sum, name='l1')
+
+
+def pred_loss(pred_probas: tf.Variable, target_probas: tf.Tensor) -> tf.Tensor:
+
+    # TODO: ALEX: DOCSTRING
+
+    return tf.square(pred_probas - target_probas)
+
+
+def range_constraint(X: Union[tf.Variable, tf.Tensor], low: Union[float, np.ndarray], high: Union[float, np.ndarray]):
+
+    # TODO: ALEX: TBD: SHOULD WE DO ERROR HANDLING FOR THIS FUNCTION (E.G., LOW < HIGH)
+
+    # TODO: ALEX: ADD DOCSTRING,
+    # TODO: ALEX: CROSS REFERENCE INIT DOC FOR MIN AND MAX, MAKE SURE THEY DEFINE SHAPES FOR MIN/MAX
+
+    return tf.clip_by_value(X, clip_value_min=low, clip_value_max=high)
+
```

Comment:
To allow easy constraint of specific parameter values (e.g. can't change the age of a person but it's possible to change any other feature) clipping fn should be done in a user-friendly way by feature.